near-plugins/tests/full_access_key_fallback.rs

```diff
@@ -0,0 +1,171 @@
+        .unwrap()
+}
+
+/// Converts a `near_sdk::PublicKey` to a `workspaces::types::PublicKey`.
+fn pk_sdk_to_workspaces(public_key: near_sdk::PublicKey) -> PublicKey {
+    #[derive(Deserialize)]
+    struct Wrapper {
+        public_key: PublicKey,
+    }
+
+    let ser = json!({ "public_key": public_key });
+    from_value::<Wrapper>(ser).unwrap().public_key
+}
+
+/// Allows spinning up a setup for testing the contract in [`PROJECT_PATH`] and bundles related
+/// resources.
+struct Setup {
+    /// Instance of the deployed contract.
+    contract: Contract,
+    /// Wrapper around the deployed contract that facilitates interacting with methods provided by
+    /// the `FullAccessKeyFallback` plugin.
+    fa_key_fallback_contract: FullAccessKeyFallbackContract,
+    /// A newly created account without any `Ownable` permissions.
+    unauth_account: Account,
+}
+
+impl Setup {
+    /// Deploys and initializes the contract in [`PROJECT_PATH`] and returns a new `Setup`.
+    ///
+    /// The `owner` parameter is passed on to the contract's constructor, allowing to optionally set
+    /// the owner during initialization.
+    async fn new(worker: Worker<Sandbox>, owner: Option<AccountId>) -> anyhow::Result<Self> {
+        // Compile and deploy the contract.
+        let wasm =
+            common::repo::compile_project(Path::new(PROJECT_PATH), "full_access_key_fallback")
+                .await?;
+        let contract = worker.dev_deploy(&wasm).await?;
+        let fa_key_fallback_contract = FullAccessKeyFallbackContract::new(contract.clone());
+
+        // Call the contract's constructor.
+        contract
+            .call("new")
+            .args_json(json!({
+                "owner": owner,
+            }))
+            .max_gas()
+            .transact()
+            .await?
+            .into_result()?;
+
+        let unauth_account = worker.dev_create_account().await?;
+        Ok(Self {
+            contract,
+            fa_key_fallback_contract,
+            unauth_account,
+        })
+    }
+
+    /// Asserts the contract's access keys are:
+    ///
+    /// - the contracts own key followed by
+    /// - the keys specified in `keys`
+    ///
+    /// Moreover, it asserts that all access keys have `FullAccess` permission.
+    async fn assert_full_access_keys(&self, keys: &[PublicKey]) {
+        // Assert the number of keys.
+        let access_keys = self
+            .contract
+            .view_access_keys()
+            .await
+            .expect("Should view access keys");
+        assert_eq!(
+            access_keys.len(),
+            keys.len() + 1, // + 1 for the contract's key
+        );
+
+        // Assert the `access_keys` are the contract's key followed by `keys` (all full access).
+        let contract_key = self.contract.as_account().secret_key().public_key();
+        let expected_keys = iter::once(&contract_key).chain(keys.iter());
+        for (i, expected_key) in expected_keys.into_iter().enumerate() {
```

Comment:
Seems a little fragile to me that we rely on the specific order of the keys in this test. Is there a specific order that is expected from `view_access_keys`? If not then I think it would be better to compare these as sets instead.

Reply:
As [view_access_keys](https://docs.rs/workspaces/latest/workspaces/struct.Contract.html#method.view_access_keys) returns a list I was _assuming_ the returned keys are ordered. Since that’s not specified in the docs, you’re right, it’s not a solid approach.

It’s refactored to not rely on the order in 118de52. Currently it’s not possible to use a (`std`) set since `workspaces::types::PublicKey` doesn’t implement neither `Hash` nor `Ord`. I’ve opened [this issue](https://github.com/near/workspaces-rs/issues/263).